### PR TITLE
[XLA:GPU] Fix race condition in nccl collectives

### PR DESCRIPTION
**Pull request by ymwangg**
This PR fixed the race condition in nccl_all_reduce due to async_comm_stream.

To reproduce this issue, clone [pytorch/xla](https://github.com/pytorch/xla) and apply the following patch:
```diff
diff --git a/test/test_train_mp_imagenet_amp.py b/test/test_train_mp_imagenet_amp.py
index 734d9c69..9d9a5b4c 100644
--- a/test/test_train_mp_imagenet_amp.py
+++ b/test/test_train_mp_imagenet_amp.py
@@ -224,18 +224,16 @@ def train_imagenet():
   def train_loop_fn(loader, epoch):
     tracker = xm.RateTracker()
     model.train()
+    zero = torch.tensor(0.0).to(device)
     for step, (data, target) in enumerate(loader):
       optimizer.zero_grad()
       if FLAGS.amp:
-        with autocast():
-          output = model(data)
-          loss = loss_fn(output, target)
-
-        scaler.scale(loss).backward()
+        output = model(data)
+        loss = loss_fn(output, target)
+        loss.backward()
         gradients = xm._fetch_gradients(optimizer)
         xm.all_reduce('sum', gradients, scale=1.0 / xm.xrt_world_size())
-        scaler.step(optimizer)
-        scaler.update()
+        optimizer.step(found_inf=zero)
       else:
         output = model(data)
         loss = loss_fn(output, target)
```

Train resnet-50 using the following command:
```
GPU_NUM_DEVICES=8 python test/test_train_mp_imagenet_amp.py --amp --use_syncfree_optim --datadir /home/ubuntu/dataset/imagenet/ILSVRC/Data/CLS-LOC/ 
```

Without this patch, there should be NaN loss within a few hundred steps.
With this patch, the model can achieve ~10% accuracy after running 1 epoch.
Tested on 8 Nvidia V100 GPU.

We noticed this issue a long time ago (https://github.com/pytorch/xla/issues/3200). I happened to the above training script can reliably reproduce this bug while debugging DETR model and syncfree optimizers.

Note I need to revert the following commit to avoid a duplicate request_id error (issue to be filed separately, but seems irrelevant to this PR).
```
commit 0de6ecda97e261528b51709c11a4e7e22a39ca33
Author: A. Unique TensorFlower <gardener@tensorflow.org>
Date:   Wed Sep 28 14:48:25 2022 -0700

    Use thread local random generator to generate request id for RpcRecvTensorCall since using a singleton random generator can cause lock contention in large scale distributed training.
    
    PiperOrigin-RevId: 477557876
```

### CODE REVIEW

**Comment by cheshire** [comment]
@gcforster could you take a look?

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 1]
```diff
@@ -366,6 +366,17 @@ Status ExecuteThunks(const std::string& module_name,
         *run_options, buffer_allocations, main_stream,
         async_comms_stream.ok() ? async_comms_stream->get() : nullptr};
     TF_RETURN_IF_ERROR(thunk->ExecuteOnStream(thunk_params));
+
+    // GPU multi-stream is currently not supported. We need to make
+    // exeuctions on async_comms_stream blocking to avoid race condition.
+    if (async_comms_stream.ok() && NeedsAsyncCommsStream(*thunk)) {
+      Status block_status = async_comms_stream->get()->BlockHostUntilDone();
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc:373]: Just `TF_RETURN_IF_ERROR(async_comms_stream->get()->BlockHostUntilDone())`? We can pipe extra info to that, e.g. `TF_RETURN_IF_ERROR(...) << "extra logging"`

**Review by cheshire — CHANGES_REQUESTED** [review]
Is it possible to test this at all?

**Comment by ymwangg** [comment]
> Is it possible to test this at all?

Is it ok to test it in the downstream [pytorch/xla](https://github.com/pytorch/xla) CI? I found this [test case](https://github.com/pytorch/xla/blob/master/test/test_ddp.py) in pytorch/xla CI can capture this bug. I can enable it on GPU once pytorch/xla update its TF pin.

I found it's difficult to create a similar test case on TF side because currently all_reduce and optimizer step cannot be compiled (mentioned in https://github.com/tensorflow/tensorflow/issues/45940#issuecomment-759746281, I did a quick test and this statement still looks true).

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc (hunk 1 above)
Comment (ymwangg) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: Changed.

**Comment by cheshire** [comment]
> Is it ok to test it in the downstream [pytorch/xla](https://github.com/pytorch/xla) CI?

Ideally, could we test it in XLA itself, without a framework to depend on.

https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/tests/collective_ops_test.cc;l=1?q=f:collective_ops_test&sq= tests collectives inside XLA, is it possible to do something similar?

**Comment by ymwangg** [comment]
Thanks for the info. I'll see if I can figure out a cpp test case.

**Comment by ymwangg** [comment]
@cheshire Upon second look, I found the race condition is not between the `main_stream` and `async_comms_stream` and these two streams are synchronized by an stream event [here](https://github.com/tensorflow/tensorflow/blob/9f828f5ef1f13b21b60f83294c6ee91c9a4b49f6/tensorflow/compiler/xla/service/gpu/nccl_all_reduce_thunk.cc#L332). The race condition is actually between `async_comms_stream` and async buffer deallocation [here](https://github.com/tensorflow/tensorflow/blob/813bbd28c16c70c5956fe2ba8d37fbf3c6133b73/tensorflow/compiler/xla/service/gpu/gpu_executable.cc#L878). So it's sufficient to only call `async_comm_stream->BlockHostUntilDone()` once before calling `buffer_allocations.TearDown()`. Though the previous fix (calling BlockHostUntilDone() multiple times) works but it's suboptimal. This bug only affects executions using allocators with `AllowsAsynchronousDeallocation==True` such as `GPUBFCAllocator`, which is used in torch_xla GPU by default.

I checked the example test cases in `collective_ops_test.cc` and it looks like `HloRunner` only supports the default allocator (backend().memory_allocator()). Is it possible to run these tests with `GPUBFCAllocator`?

**Comment by cheshire** [comment]
> only supports the default allocator (backend().memory_allocator()). Is it possible to run these tests with GPUBFCAllocator

Is it possible to pipe through a method which supplies a custom allocator? TF runtime does pipe it through, so the API must be there?

**Comment by ymwangg** [comment]
I'm hitting OOM error when simply replacing `backend().memory_allocator()` with `GPUBFCAllocator` in `HloRunner`. So maybe it's the user's responsibility to free up the allocated memory when using `BFCAllocator`. Additionally, upon looking into the codebase, I only found [PjRT](https://github.com/tensorflow/tensorflow/blob/1cb0c5b850657ae1362a241fabb16253336dd8c3/tensorflow/compiler/xla/pjrt/gpu/gpu_helpers.cc#L74) and [XRT](https://github.com/tensorflow/tensorflow/blob/1cb0c5b850657ae1362a241fabb16253336dd8c3/tensorflow/compiler/xrt/xrt_device.cc#L120) uses `GPUBFCAllocator`.

I'll see if I can add a test case using PjRT or resolve the OOM error in HloRunner.

**Comment by cheshire** [comment]
Thanks a lot for looking into it ! Sorry for being pedantic over tests, but then it would be hard to know we won't regress the added fix.

@hawkinsp might also help with passing proper alloocator through runtime.

**Comment by gbaned** [comment]
Hi @ymwangg Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @ymwangg I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! thank you for your contribution.

**Comment by ymwangg** [comment]
Can someone from Google take this over? I tried several test cases but non of them can capture this bug in tensorflow.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 2]
```diff
@@ -367,6 +367,18 @@ Status ExecuteThunks(const std::string& module_name,
...
+  // bfc_allocator).
+  if (async_comms_stream.ok() &&
+      run_options->allocator()->AllowsAsynchronousDeallocation()) {
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(
+        async_comms_stream->get()->BlockHostUntilDone(),
+        absl::StrFormat("Failed to complete all kernels launched on stream %p",
+                        async_comms_stream->get()));
+  }
+
   return MaybeSyncAndProfile(run_options, start_micros,
```
Comment (jurahul) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc:382]: Isn't the async comm stream always synchronized with the main stream (as a part of NcclCollectiveDoneThunk)? In which case at the end of the execution, blocking host on the main stream (if `block_host_until_done` is true) should be sufficient?

Comment (ymwangg) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc:382]: Yes, sync on main_stream here also works. The reason I'm using async_comms_stream is due to its lesser scope given the race condition is between async_comms_stream and buffer tear down.

**Comment by gbaned** [comment]
Hi @ymwangg Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @ymwangg Can you please resolve conflicts? Thank you!
